File: skdltn210/week11/63.js
/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */

var uniquePathsWithObstacles = function (obstacleGrid) {
  let dp = Array(obstacleGrid.length)
    .fill(0)
    .map(() => Array(obstacleGrid[0].length).fill(0));

  for (let i = 0; i < obstacleGrid[0].length; i++) {
    if (obstacleGrid[0][i] === 1) {
      break;
    }
    dp[0][i] = 1;
  }

  for (let i = 0; i < obstacleGrid.length; i++) {
    if (obstacleGrid[i][0] === 1) {
      break;
    }
    dp[i][0] = 1;
  }

  for (let i = 1; i < obstacleGrid.length; i++) {
    for (let j = 1; j < obstacleGrid[0].length; j++) {
      if (obstacleGrid[i][j] === 1) {
        dp[i][j] = 0;
      } else {
        dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
      }
    }
  }
  return dp[obstacleGrid.length - 1][obstacleGrid[0].length - 1];
};

File: simjaesung/week11/63.java
class Solution {
    public int[][] dp;
    public int n, m;

    public boolean inRange(int x, int y){
        return x >= 0 && x < m && y >= 0 && y < n;
    }

    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        if(obstacleGrid[0][0] == 1) return 0;
        m = obstacleGrid.length;
        n = obstacleGrid[0].length;
        dp = new int[m][n];
        dp[0][0] = 1;

        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(obstacleGrid[i][j] == 1) continue;
                if(inRange(i-1,j)) dp[i][j] += dp[i-1][j];
                if(inRange(i,j-1)) dp[i][j] += dp[i][j-1];
            }
        }

        return dp[m-1][n-1];
    }
}